src/utils/userLocation.ts

```diff
@@ -0,0 +1,120 @@
+import firebase, { UserInfo } from "firebase/app";
+import { isArray, isEmpty } from "lodash";
+
+import { AnyRoom } from "types/rooms";
+import { AnyVenue, VenueEvent } from "types/venues";
+import { User } from "types/User";
+
+import { updateUserProfile } from "pages/Account/helpers";
+import { useInterval } from "hooks/useInterval";
+
+import { WithId } from "./id";
+import { getCurrentTimeInMilliseconds } from "./time";
+import { openRoomUrl, openUrl, venueInsideUrl } from "./url";
+
+const LOCATION_INCREMENT_SECONDS = 10;
+const LOCATION_INCREMENT_MS = LOCATION_INCREMENT_SECONDS * 1000;
+
+type LocationData = Record<string, number>;
+
+export const updateLocationData = (
+  userId: string,
+  newLocationData: LocationData = {}
+) => {
+  updateUserProfile(userId, {
+    lastSeenAt: getCurrentTimeInMilliseconds(),
+    lastSeenIn: newLocationData,
+  });
+};
+
+export interface TrackLocationProps {
+  userId: string;
+  locationName: string;
+  lastSeenIn?: LocationData;
+}
+
+const trackLocationEntered = ({ userId, locationName }: TrackLocationProps) => {
+  updateLocationData(userId, {
+    [locationName]: getCurrentTimeInMilliseconds(),
+  });
+};
+
+export interface BaseEnterRoomWithCountingProps {
+  user: User;
+  profile?: User;
+  venue: WithId<AnyVenue>;
+}
+
+export interface EnterRoomWithCounting extends BaseEnterRoomWithCountingProps {
+  room?: AnyRoom;
+}
+
+type EnterVenueRoomProps = {
+  userId: string;
+  venueId: string;
+  venueName: string;
+};
+
+export const enterVenueRoom = ({
+  userId,
+  venueName,
+  venueId,
+}: EnterVenueRoomProps) => {
+  trackLocationEntered({
+    userId,
+    locationName: venueName,
+  });
+  openUrl(venueInsideUrl(venueId));
+};
+
+type EnterExternalRoomProps = {
+  userId: string;
+  room: AnyRoom;
+};
+
+export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
+  trackLocationEntered({
+    userId,
+    locationName: `external/${room.title}`,
+  });
```

Comment:
This `external/` seems like a new concept being introduced here. Would be good to know what your thinking is around it/how it will work/be used/etc.